### PR TITLE
ENH: Boundary conditions for new CubicSpline

### PR DESCRIPTION
**Pull request by matmel**
Following the discussion in PR #5653, and as suggested by @ev-br and @nmayorov I'm opening this new pull request to track my changes related to the implementation of `bc_type` argument for the new great `CubicSpline` class implemented by Nikolay.

Note that I may not have too much time anymore for the two following weeks to work on that, but I definitely intend to address the issues raised by Evgeni and @sturlamolden in the referenced pull request.

### CODE REVIEW

**Comment by ev-br** [comment]
Great, looking forward for your getting back to it! Meanwhile, I'm adding it to the milestone 0.18 and tagging it as a needs-work tag, just to communicate my assessment of the status and intentions (please speak up if you disagree with either).

**Comment by ev-br** [comment]
To clarify the state of the PR: this is branched off gh-5653, so only commits starting from https://github.com/matmel/scipy/commit/588c9baddeb7cde133b7a3476e15506fb6eb11eb are relevant here.

There is prior discussion of what is now this PR starting from https://github.com/scipy/scipy/pull/5653#issuecomment-172398734

BTW, the CI failure is unrelated.

**Comment by ev-br** [comment]
@nmayorov would you be interested in reviewing this PR?

File: scipy/interpolate/_cubic.py [hunk 1]
```diff
@@ -0,0 +1,754 @@
...
+                b2 = np.zeros(b1.shape, b1.dtype)
+                b2[0] = -a_0_m1
+                b2[-1] = -a_m2_m1
+
+                # s1 and s2 are the solutions of (n-2, n-2) system
+                s1 = solve_banded((1, 1), Ac, b1, overwrite_ab=False,
+                                 overwrite_b=False, check_finite=False)
+
+                s2 = solve_banded((1, 1), Ac, b2, overwrite_ab=False,
+                                 overwrite_b=False, check_finite=False)
```
Comment (ev-br) [inline scipy/interpolate/_cubic.py]: solve_banded should be able to handle an (m, 2) right-hand side

**Comment by nmayorov** [comment]
General suggestion about testing. All boundary conditions should be tested on some non-trivial input (`test_general` for example) in the following way:

1. Test spline continuity properties. I wrote a function for that.
2. Test that boundary conditions hold. 

Value comparison from `test_general` can be removed.

**Comment by nmayorov** [comment]
Apart from a few notes (subject to discussion), I found this implementation to be quite excellent, complete and easy to follow. But the tests must be improved.

**Comment by ev-br** [comment]
Should be rebase on master now that gh-5653 is in.

**Comment by auvipy** [comment]
@matmel do u have time to re base it?

**Comment by nmayorov** [comment]
If @matmel doesn't mind I can try to finish this PR, especially if it's necessary for 0.18.0 release.

**Comment by ev-br** [comment]
Continued in and superseded by https://github.com/scipy/scipy/pull/6181, so closing.

**Pull request closed without merging**